cluster resource fetcher for IBM Cloud

tmishina opened the pull request:
- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](../blob/master/DCO1.1.txt)


## What

Add functionality of fetching resources from IBM Cloud Kubernetes clusters.

## Why

Kubernetes resources (e.g., output of `kubectl get pod`) can be used as evidence. For example, spec of Pod, custom resource of an operator, and ConfigMap shows whether applications (pod) and kubernetes infrastructure (operator) run with correct (expected) configuration. Managed clusters have their own mechanism to manage multiple clusters, and therefore a specific fetcher for each managed cluster is required. This PR contains a fetcher for IBM Cloud clusters.

## How

- add Kubernetes resource fetcher for IBM Cloud clusters to `ibm_cloud` category
- add a package `PyYAML` to `setup.cfg` to parse configuration of IBM Cloud clusters
- modify `arboretum/kubernetes/README.md` to add link to the cluster resource fetcher section in `arboretum/ibm_cloud/README.md`


## Test

- fetching resources from IBM Cloud clusters (both IBM Cloud Kubernetes Service and Red Hat OpenShift Kubernetes Service)

## Context

- issue: #9 
- related PR: #32 - this PR has been spun off from #32

tmishina committed:
feat: IBM Cloud cluster resource fetcher

tmishina committed:
docs: add link to IBM Cloud cluster resource fetcher

tmishina commented:
I'd like to get reviewers' comments for the following items. Thanks in advance.

1. Are these handling (raising a `RuntimeError` when a fatal error happens) suitable?
    - https://github.com/ComplianceAsCode/auditree-arboretum/pull/54/files#diff-bcd498e6b9c862fab4f20772411039245ac766ae4cab4186d5c9285b8835064aR75-R85
    - https://github.com/ComplianceAsCode/auditree-arboretum/pull/54/files#diff-bcd498e6b9c862fab4f20772411039245ac766ae4cab4186d5c9285b8835064aR110-R114
2. Can I fixup two commits in this PR? I split it because they modify different category (`ibm_cloud` and `kubernetes`), but I know that PR should contain one commit.

I will bump version in and add a line to CHANGES.md when we are ready to merge this PR.

alfinkel commented:
The PR can have as many commits as you want.  The repo is set to squash all commits to one as part of a PR merge anyway.

alfinkel commented:
If the fetcher is going to error anyway then I wouldn't bother raising a RuntimeError.  Historically we've been pretty sparse with raising errors in fetchers and checks if the result produces an error anyway.  If processing won't stop for some reason and you want it to then raise the error otherwise, I would let Python raise the error which should hopefully produce a useful stack trace.

tmishina committed:
chore: remove empty line

tmishina committed:
chore: remove raising exception to simplify code

tmishina commented:
@alfinkel Thank you for your comment, I checked "raise" lines and removed them because error and its cause can be found at the error message that Python runtime generates.  Also, I rebased the branch to v0.11.1.

alfinkel commented on arboretum/ibm_cloud/README.md:
```diff
@@ -62,6 +62,49 @@ to include the fetchers and checks from this library in your downstream project.
+### Cluster Resource
+
+* Class: [ClusterResourceFetcher][fetch-ibm-cloud-cluster-resource]
+* Purpose: Write the resources of IBM Cloud Kubernetes clusters to the evidence locker.
```
How does this fetcher differ from the `kubernetes` fetcher?

tmishina replied:
I added description about the difference in line 69.
https://github.com/ComplianceAsCode/auditree-arboretum/pull/54/files#diff-5932d4f823fe9fa3a1b95d08c50cd53ef4f485aa7e3adc41da52628ceda1465dR69

alfinkel replied:
Seems to me like the purpose of this fetcher is complementary to the stand-alone cluster resource fetcher. 
 If so, then shouldn't the write up be the complement to that one?  Something like:

```
Purpose: Write the resources of **managed** Kubernetes clusters to the evidence locker. 
NOTE: Do not use this fetcher for stand-alone clusters. For Kubernetes stand-alone clusters, 
use the [Kubernetes cluster resource fetcher][fetch-kube-cluster-resource].
```

alfinkel commented on arboretum/ibm_cloud/README.md:
```diff
@@ -62,6 +62,49 @@ to include the fetchers and checks from this library in your downstream project.
+
+* Class: [ClusterResourceFetcher][fetch-ibm-cloud-cluster-resource]
+* Purpose: Write the resources of IBM Cloud Kubernetes clusters to the evidence locker.
+* Behavior: Retrieve the resources of IBM Cloud Kubernetes clusters listed by [cluster list fetcher][fetch-cluster-list]. For IBM Cloud Kubernetes Service (IKS) clusters, the fetcher [downloads cluster config via IBM Cloud API][ibm-cloud-download-config] and extract the access token from the config. For IBM Cloud Red Hat Kubernetes Service (ROKS) clusters, the fetcher retrieves the access token u ...
```
Is it relevant to call out how the access token is retrieved? This behavior doesn't describe anything having to do with the fetcher behavior but more about how access seems to be acquired to perform the fetch.

tmishina replied:
I removed the description of internal behavior and move it to code comments of [IKS cluster function](https://github.com/ComplianceAsCode/auditree-arboretum/pull/54/files#diff-bcd498e6b9c862fab4f20772411039245ac766ae4cab4186d5c9285b8835064aR66-R67) and [ROKS cluster function](https://github.com/ComplianceAsCode/auditree-arboretum/pull/54/files#diff-bcd498e6b9c862fab4f20772411039245ac766ae4cab4186d5c9285b8835064aR87-R88).

alfinkel commented on arboretum/ibm_cloud/README.md:
```diff
@@ -62,6 +62,49 @@ to include the fetchers and checks from this library in your downstream project.
+  * `org.ibm_cloud.cluster_resources.target_resource_types`
+    * Optional
+    * List of resource types as strings
+    * See the document of the [Kubernetes resource fetcher][fetch-kube-cluster-resource] for details.
```
?

This link is sending you to the code for the Kubernetes cluster resource fetcher.  It's not clear why.

tmishina replied:
I changed the link target to the description in the README file of the Kubernetes category.  Does this make sense for you?

alfinkel replied:
Maybe I'm missing something but why are you repeatedly referencing the kubernetes resource fetcher?  As a user, I'd want to be able to go to this README, understand what the fetcher will do for me and how I can configure it.  It seems like you're just not wanting to replicate the content from that other README, is that it?  If so then I think that's wrong.

alfinkel replied:
The config of the kubernetes fetcher isn't even the same as this one.  This is super confusing.

If target_resource_types are optional then the reader needs to know what the default is if the config is not provided.

alfinkel commented on arboretum/ibm_cloud/README.md:
````diff
@@ -62,6 +62,49 @@ to include the fetchers and checks from this library in your downstream project.
+  ```
+
+* Required credentials:
+  * `ibm_cloud` credentials with read/view permissions are needed for this fetcher to successfully retrieve the evidence.  See `Required credentials` of the [cluster list fetcher][fetch-cluster-list].
````
You should reiterate it here rather than forcing someone to have to jump around for this info.

tmishina replied:
I copied the description here.

alfinkel commented on arboretum/ibm_cloud/README.md:
````diff
@@ -62,6 +62,49 @@ to include the fetchers and checks from this library in your downstream project.
+* Import statement:
+
+   ```python
+   from arboretum.ibm_cloud.fetchers.fetch_cluster_resource import ClusterResourceFetcher
````
`ClusterResourceFetcher` is also the name of the `kubernetes` version of this fetcher.  If added to the same import file this will cause a name collision and it won't be apparent why one of the fetchers (the first one) is not running.  Since the kubernetes version is already in place and we cannot really change the name of the class there, it would be a good idea to rename this one.  Maybe something like `ICClusterResourceFetcher`?

tmishina replied:
I renamed it to as you suggested (`ICClusterResourceFetcher`).

alfinkel reviewed (commented):
It would be good for you to clarify the behavior of this fetcher in the README and how it differs from the kubernetes fetcher of the same name.  It would also be good for the review if it were clear why you appear to be writing files to the root $TMPDIR and why you're also zipping files?

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,151 @@
+    @store_raw_evidence('ibm_cloud/cluster_resources.json')
+    def fetch_cluster_resource(self):
+        """Fetch cluster resources."""
+        with evidences(self.locker, 'raw/ibm_cloud/cluster_list.json') as ev:
```
`evidences` is the context manager for checks not fetchers.  You can't do this.  I'm  not really sure how it even works.

alfinkel replied:
I don't really even understand what's happening here.  Putting aside the misuse of the `evidences` context manager.  It appears that this fetcher depends on evidence gathered by another fetcher?  That really isn't clear in the README write up.  If that is your intent then you need to use get_evidence_dependency otherwise this fetcher will not run correctly all the time.  I'm surprised that it hasn't failed for you during your testing.

tmishina replied:
I was not clear about `with evidences`; I changed this to using [get_evidence_dependency](https://github.com/ComplianceAsCode/auditree-arboretum/pull/54/files#diff-bcd498e6b9c862fab4f20772411039245ac766ae4cab4186d5c9285b8835064aR119-R122), and the test result is as follows.

```
 $ compliance  --fetch -C ../auditree-data/cluster-list-fetcher/config.ibm_cloud.multi_account_custom_rsrc.json arboretum.ibm_cloud.fetchers.fetch_cluster_resource arboretum.ibm_cloud.fetchers.fetch_cluster_list --evidence local
INFO: Creating locker in /var/folders/s8/57ct85zd407d1g6cx0dtjhqw0000gn/T/compliance...

Fetcher Primary Run

E.
======================================================================
ERROR: fetch_cluster_resource (arboretum.ibm_cloud.fetchers.fetch_cluster_resource.ICClusterResourceFetcher)
Fetch cluster resources.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/tmishina/workspace/auditree-evolve/compliance-evidence-collector/venv/lib/python3.7/site-packages/compliance/locker.py", line 456, in get_evidence
    metadata['ttl'],
TypeError: 'NoneType' object is not subscriptable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/tmishina/workspace/auditree-evolve/compliance-evidence-collector/venv/lib/python3.7/site-packages/compliance/evidence.py", line 637, in get_evidence_dependency
    evidence = locker.get_evidence(path)
  File "/Users/tmishina/workspace/auditree-evolve/compliance-evidence-collector/venv/lib/python3.7/site-packages/compliance/locker.py", line 465, in get_evidence
    f'Evidence {evidence_path} is not valid '
compliance.utils.exceptions.EvidenceNotFoundError: Evidence raw/ibm_cloud/cluster_list.json is not valid or not present within the locker

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/tmishina/workspace/auditree-evolve/compliance-evidence-collector/venv/lib/python3.7/site-packages/compliance/evidence.py", line 684, in wrapper
    return _store_wrapper(self, evidence_path, f, 'raw')
  File "/Users/tmishina/workspace/auditree-evolve/compliance-evidence-collector/venv/lib/python3.7/site-packages/compliance/evidence.py", line 840, in _store_wrapper
    content = func(self)
  File "/Users/tmishina/workspace/auditree-evolve/compliance-evidence-collector/arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py", line 120, in fetch_cluster_resource
    'raw/ibm_cloud/cluster_list.json', self.locker
  File "/Users/tmishina/workspace/auditree-evolve/compliance-evidence-collector/venv/lib/python3.7/site-packages/compliance/evidence.py", line 659, in get_evidence_dependency
    f'evidence dependency {path} is currently unavailable.'
compliance.utils.exceptions.DependencyUnavailableError: evidence dependency raw/ibm_cloud/cluster_list.json is currently unavailable.

----------------------------------------------------------------------
Ran 2 tests in 3.030s

FAILED (errors=1)

Fetcher Dependency Re-Run #1

.
----------------------------------------------------------------------
Ran 1 test in 87.288s

OK
INFO: Committing changes to local locker in /var/folders/s8/57ct85zd407d1g6cx0dtjhqw0000gn/T/compliance...

```

alfinkel reviewed (changes requested):
My suggestion here is for you to rewrite this fetcher because you don't appear to be using the framework properly.

tmishina committed:
refactor: rename fetcher class name to distinguish it from another same-name fetcher in kubernetes category

tmishina committed:
docs: change link to kubernetes cluster fetcher to point README file

tmishina committed:
docs: move internal behavior descriptions to program source code

tmishina committed:
docs: add description about difference between kubernetes resource fetcher and IBM Cloud resource fetcher

tmishina committed:
docs: reiterate description for required credentials

tmishina committed:
fix: evidence dependency handling

tmishina commented:
@alfinkel Thank you very much for your detailed review. I updated the code and the document as commented above.  Please check them and tell me if further update is required.  Thanks again.

alfinkel commented on arboretum/ibm_cloud/README.md:
```diff
@@ -62,6 +62,64 @@ to include the fetchers and checks from this library in your downstream project.
+
+* Class: [ICClusterResourceFetcher][fetch-ibm-cloud-cluster-resource]
+* Purpose: Write the resources of IBM Cloud Kubernetes clusters to the evidence locker.
+* Behavior: Retrieve tokens for IBM Cloud Kubernetes clusters listed by [IBM Cloud cluster list fetcher][fetch-cluster-list] using `api_key` in `~/.credentials`, and then retrieve specified `target_resource_types` of the clusters using the tokens. This fetcher uses one `api_key` per account which owns multiple clusters, while [Kubernetes resource fetcher][fetch-kube-cluster-resource] uses one `to ...
```
Is there a reason you're referencing the Kubernetes resource fetcher here?  Seems unnecessary.

alfinkel reviewed (changes requested):
@tmishina It appears that this fetcher is the mirror of the kubernetes resources fetcher in that this fetcher fetches resources for managed clusters while the kubernetes version fetches resources for stand-alone clusters.

This fetcher's README write up **and configuration** should be as similar as possible to that of the kubernetes version.  For example there doesn't appear to be any reason to refer to cluster resource types here as cluster_resources.target_resource_types while in the kubernetes version they are cluster_resources.types.

Please rewrite your README content (and make the necessary code changes) so that the two fetchers are more in line.  Please don't send the reader to the other fetcher for details on configuration of this fetcher.  The only thing necessary here is to reference the other fetcher as a NOTE in the "purpose" section when describing what to do for folks that want to fetch evidence for stand-alone clusters.

tmishina committed:
docs: revise descriptions for purpose and behavior

tmishina committed:
docs: now resource type config keyword is "types"

tmishina committed:
docs: describe details for "types" instead of link

tmishina committed:
docs: add example configuration for IBM Cloud cluster resource fetcher

tmishina committed:
docs: fix json key error

tmishina commented:
@alfinkel I have updated README and example configration. Could you review this? Thanks in advance.

- update [README.md](https://github.com/tmishina/auditree-arboretum/blob/9-ibm-cloud-resource-fetcher/arboretum/ibm_cloud/README.md) to apply your suggestions and comments
  - change config name `target_resource_types` to `types`
  - add config description instead of link
  - update `purpose`
- add example configuration for this fetcher to [devel.json](https://github.com/tmishina/auditree-arboretum/blob/9-ibm-cloud-resource-fetcher/devel.json)

alfinkel commented on arboretum/ibm_cloud/README.md:
```diff
@@ -62,6 +62,81 @@ to include the fetchers and checks from this library in your downstream project.
+* Configuration elements:
+  * `org.ibm_cloud.accounts`
+    * Required
+    * List of accounts (string)
```
```suggestion
    * List of accounts as strings
```

alfinkel commented on arboretum/ibm_cloud/README.md:
```diff
@@ -62,6 +62,81 @@ to include the fetchers and checks from this library in your downstream project.
+
+* Class: [ICClusterResourceFetcher][fetch-ibm-cloud-cluster-resource]
+* Purpose: Write the resources of **managed** Kubernetes clusters to the evidence locker. NOTE: Do not use this fetcher for stand-alone clusters. For Kubernetes stand-alone clusters, use the [Kubernetes cluster resource fetcher][fetch-kube-cluster-resource].
+* Behavior: Retrieve tokens for IBM Cloud Kubernetes clusters listed by [IBM Cloud cluster list fetcher][fetch-cluster-list] using `api_key` in `~/.credentials`, and then retrieve specified resources  using the tokens. TTL is set to 1 day.
```
```suggestion
* Purpose: Write the resources of **managed** Kubernetes clusters to the evidence locker.
* Behavior: Retrieve managed Kubernetes cluster resource data based on clusters gathered by the [IBM Cloud cluster list fetcher][fetch-cluster-list].  TTL is set to 1 day.
* NOTE: 
   * Do not use this fetcher for stand-alone clusters. For Kubernetes stand-alone clusters, use the [Kubernetes cluster resource fetcher][fetch-kube-cluster-resource].
   * This fetcher is dependent on evidence gathered by the [IBM Cloud cluster list fetcher][fetch-cluster-list], 
 i.e. importing the IBM Cloud cluster list fetcher is a prerequisite for the IKS cluster resource fetcher to work.
```

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,165 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Feching cluster resource from IBM Cloud."""
```
```suggestion
"""IBM Cloud cluster resource fetcher."""
```

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,165 @@
+        cluster_list_evidence = get_evidence_dependency(
+            'raw/ibm_cloud/cluster_list.json', self.locker
+        )
+        cluster_list = json.loads(cluster_list_evidence.content)
```
```suggestion
        cluster_list = cluster_list_evidence.content_as_json
```

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,165 @@
+        for account in cluster_list:
+            api_key = getattr(
+                self.config.creds['ibm_cloud'], f'{account}_api_key'
+            )
```
Does this not work?:

```suggestion
            api_key = self.config.creds.get('ibm_cloud', f'{account}_api_key')
```

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,165 @@
+            for cluster in cluster_list[account]:
+                resp = self.session().get(
+                    '/global/v1/clusters/'
+                    f'{cluster["id"]}/config'
```
```suggestion
                    f'/global/v1/clusters/{cluster["id"]}/config'
```

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,165 @@
+                    io.BytesIO(resp.content)
+                )
+            resources[account] = []
+            for cluster in cluster_list[account]:
```
It seems inefficient to step through the same exact loop twice.  Seems like if you changed `_get_iks_credentials` to accept an individual cluster config rather than the full dictionary then you could do this with a single loop?

tmishina replied:
The original code aimed to shorten communication with the IBM cloud server by reusing session. I experimented and the result (below) showed that the reuse does not affect the performance of communication, and therefore I changed the code to the single loop.

```
1. session reuse (two for loops)
   1. 1m16.341s
   2. 1m11.159s
   3. 1m8.836s
2. no session reuse (single loop)
   1. 1m10.981s
   2. 1m13.628s
   3. 1m12.679s
```

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,165 @@
+            resources[account] = []
+            for cluster in cluster_list[account]:
+                if cluster['type'] == 'kubernetes':
+                    cluster_token, ca_cert = self._get_iks_credentials(
+                        cluster, cluster_config)
```
```suggestion
                    cluster_token, ca_cert = self._get_iks_credentials(
                        cluster, cluster_config
                    )
```

tmishina replied:
this code is now outdated (see the comment above).

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,165 @@
+                    cluster_token, ca_cert = self._get_iks_credentials(
+                        cluster, cluster_config)
+                elif cluster['type'] == 'openshift':
+                    cluster_token, ca_cert = self._get_roks_credentials(
+                        cluster, api_key)
```
```suggestion
                    cluster_token, ca_cert = self._get_roks_credentials(
                        cluster, api_key
                    )
```

alfinkel commented on arboretum/kubernetes/README.md:
```diff
@@ -21,7 +21,9 @@ to include the fetchers and checks from this library in your downstream project.
 * Class: [ClusterResourceFetcher][fetch-cluster-resource]
 * Purpose: Write the resources of **stand-alone** Kubernetes clusters to the
 evidence locker.  **NOTE:** Do not use this fetcher for managed clusters.
-Instead use the [IBM Cloud cluster list fetcher][ibm-cloud-cluster-list-fetcher].
+For IBM Cloud clusters, use the
+[IBM Cloud cluster list fetcher][ibm-cloud-cluster-list-fetcher] and the
+[IBM Cloud cluster resource fetcher][ibm-cloud-cluster-resource-fetcher].
 * Behavior: Retrieve stand-alone Kubernetes cluster resource data for the provided
 list of clusters.  TTL is set to 1 day.
```
```
* Purpose: Write the resources of **stand-alone** Kubernetes clusters to the evidence locker.
* Behavior: Retrieve stand-alone Kubernetes cluster resource data for the provided
list of clusters.  TTL is set to 1 day.
* NOTE: Do not use this fetcher for managed clusters. For IBM Cloud managed clusters, use the [IBM Cloud cluster resource fetcher][ibm-cloud-cluster-resource-fetcher].
```

tmishina replied:
applied the suggestion.

alfinkel commented on setup.cfg:
```diff
@@ -23,6 +23,7 @@ packages = find:
 install_requires =
     auditree-framework>=1.2.3
     auditree-harvest>=1.0.0
+    PyYAML
```
```suggestion
    pyyaml<5.4
```

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,165 @@
+                    f'{cluster["id"]}/config'
+                )
+                resp.raise_for_status()
+                cluster_config[cluster['name']] = zipfile.ZipFile(
+                    io.BytesIO(resp.content)
+                )
```
What is the purpose of compressing the content here and attaching that content as a value in this dictionary?

tmishina replied:
This line just create a ZipFile object with zipfile bytestream (`io.BytesIO(resp.content)`) given by the server, and does not perform compression/decompression. The zipfile object is extracted in [_get_iks_credentials()](https://github.com/ComplianceAsCode/auditree-arboretum/pull/54/files#diff-bcd498e6b9c862fab4f20772411039245ac766ae4cab4186d5c9285b8835064aR76).

alfinkel reviewed (changes requested):
To match the rest of the repo and our coding standards, please move your private methods to the bottom of your fetcher class.

tmishina committed:
docs: update config value type description

Co-authored-by: Al Finkelstein <al.finkelstein@ibm.com>

tmishina committed:
docs: revise and reorder descriptions

Co-authored-by: Al Finkelstein <al.finkelstein@ibm.com>

tmishina committed:
docs: update docstring

Co-authored-by: Al Finkelstein <al.finkelstein@ibm.com>

tmishina committed:
chore: use built-in function to read evidence as JSON

Co-authored-by: Al Finkelstein <al.finkelstein@ibm.com>

tmishina committed:
chore: simplify f-string

Co-authored-by: Al Finkelstein <al.finkelstein@ibm.com>

tmishina committed:
feat: lock PyYAML version (<5.4)

Co-authored-by: Al Finkelstein <al.finkelstein@ibm.com>

tmishina committed:
chore: use get() instead of getattr()

Co-authored-by: Al Finkelstein <al.finkelstein@ibm.com>

alfinkel commented:
It looks like your build is failing due to the code formatter.  You should have the pre-commit hook set up locally to catch such things before they get here.  In any case, it looks like this https://github.com/ComplianceAsCode/auditree-arboretum/pull/54/commits/31f04141be7d64ac2921c8532c8e94b2055dd0e2#diff-bcd498e6b9c862fab4f20772411039245ac766ae4cab4186d5c9285b8835064aR140-R142 is the culprit.  I ran the formatter locally and the suggested change that it makes is terrible IMO.  Here's what I would do instead:

```
                url = f'/global/v1/clusters/{cluster["id"]}/config'
                resp = self.session().get(url)
```

This will keep things easy to read while keeping the formatter happy.

tmishina commented:
@alfinkel sorry for bothering, I'm still working using "commit suggestion" feature in addition to the local change. The "commit suggestion" feature automatically triggers test while I haven't complete entire fix process including code formatting. I will mention to you when I complete the fix, thanks.

alfinkel commented:
@tmishina no problem at all.  I just figured that I'd chime in with a better suggested change to the code than what the code formatter is making at the moment.

tmishina committed:
refactor: simplify loop

The original code aimed to reuse session to shorten communication with
the IBM Cloud server, but according to an experiment the session reuse
did not significantly improves performance, and therefore simplicity
would be more important.

tmishina committed:
chore: format code

tmishina committed:
refactor: code format

tmishina commented:
@alfinkel I've finished updating the code and the readme. Please review the changes.  Thank you so much for your support.
I have done `git rebase main`; bumping the version string remains.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,154 @@
+            p = pathlib.Path(name)
+            if p.name.startswith('kube-config'):
+                kubeconfig = yaml.safe_load(cluster_config.read(name))
+                cluster_token = kubeconfig['users'][0]['user'][
+                    'auth-provider']['config']['id-token']
```
Please fix this.  Why not:

```suggestion
                usr = kubeconfig['users'][0]['user']
                cluster_token = usr['auth-provider']['config']['id-token']
```

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,154 @@
+        """Cleanup class."""
+        cls.tempdir.cleanup()
+
+    def _get_iks_credentials(self, cluster_config):
```
As was [requested earlier](https://github.com/ComplianceAsCode/auditree-arboretum/pull/54#pullrequestreview-647125088).  Please move this private method to the bottom of the module.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,154 @@
+                    'auth-provider']['config']['id-token']
+            if p.name.endswith('.pem'):
+                cluster_config.extract(name, path=self.tempdir.name)
+                ca_cert_filepath = os.path.join(self.tempdir.name, name)
```
Use pathlib instead of os.path.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,154 @@
+                ca_cert_filepath = os.path.join(self.tempdir.name, name)
+        return cluster_token, ca_cert_filepath
+
+    def _get_roks_credentials(self, cluster, api_key):
```
As was [requested earlier](https://github.com/ComplianceAsCode/auditree-arboretum/pull/54#pullrequestreview-647125088).  Please move this private method to the bottom of the module.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,154 @@
+            headers={'X-CSRF-Token': 'a'},
+            allow_redirects=False
+        )
+        location = resp.headers['Location']
+        keyword = 'access_token='
+        start = location.find(keyword)
+        end = location.find('&', start)
+        cluster_token = location[start + len(keyword):end]
```
Seems like this can be simplified as:

```suggestion
        location = resp.headers['Location']
        cluster_token = location.split('access_token=', 1)[1].split('&', 1)[0]
```

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,154 @@
+        start = location.find(keyword)
+        end = location.find('&', start)
+        cluster_token = location[start + len(keyword):end]
+        s.headers.update = ({'X-CSRF-Token': None})
```
I don't see the point here.  Why update session header after the fact?

tmishina replied:
I did test and confirmed that L108 is not required. Thank you for suggestion.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,154 @@
+        cluster_token = location[start + len(keyword):end]
+        s.headers.update = ({'X-CSRF-Token': None})
+
+        return cluster_token, None
```
Returning `None` here seems hacky.  If your ca cert is `None` in the case of roks creds then set it to `None` in the caller rather than here.

tmishina replied:
I changed as you suggested.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,154 @@
+                resp.raise_for_status()
+                cluster_config = zipfile.ZipFile(io.BytesIO(resp.content))
+                if cluster['type'] == 'kubernetes':
+                    cluster_token, ca_cert = self._get_iks_credentials(
+                        cluster_config
+                        )
```
Please fix this.

```suggestion
                    cluster_token, ca_cert = self._get_iks_credentials(
                        cluster_config
                    )
```

tmishina replied:
@alfinkel currently neither `make code-lint` nor `make code-format` checks this formatting.
I tried to change the behaviours and I found that changing `.pre-commit-config.yaml` as below changes the behaviour of `make code-lint` to make warning for this.

```
--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -12,7 +12,9 @@ repos:
     -   id: flake8
         args: [
           --extend-ignore,
-          "P1,C812,C813,C814,C815,C816,W503"
+          "P1,C812,C813,C814,C815,C816,W503",
+          --select,
+          "E123"
         ]
         additional_dependencies: [
           flake8-2020,
```

Unfortunately I can't find how to change the behaviour of `make code-format` (yapf); `align_closing_bracket_with_visual_indent = false` in `./yapf-config` does not work as expected.

However I think it is better to use these automatic checks than manual formatting.  Could you help me find way to change the behaviour of yapf?

alfinkel replied:
We don't want to change the linter's or the formatter's behavior.  I just want you to fix your indentation.

```suggestion
                    cluster_token, ca_cert = self._get_iks_credentials(
                        cluster_config
                    )
```

tmishina replied:
got it, I've just fixed the format of the code.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,154 @@
+                        cluster_config
+                        )
+                elif cluster['type'] == 'openshift':
+                    cluster_token, ca_cert = self._get_roks_credentials(
+                        cluster, api_key
+                        )
```
Please fix this.  See above.

alfinkel commented on devel.json:
```diff
@@ -86,7 +86,10 @@
     "ibm_cloud": {
-      "accounts": ["my_ic_account_one", "my_ic_account_two"]
+      "accounts": ["my_ic_account_one", "my_ic_account_two"],
+      "cluster_resources": {
+        "types": ["pods"]
+        }
```
Fix the indentation here to line things up correctly.

```suggestion
      "cluster_resources": {
        "types": ["pods"]
       }
```

alfinkel reviewed (changes requested):
If you apply all suggestions then hopefully this is the last round of changes.  Feel free to bump version, add content to CHANGES.md and squash commits.

tmishina committed:
chore: improve code readability

Co-authored-by: Al Finkelstein <al.finkelstein@ibm.com>

tmishina committed:
chore: simplify code

Co-authored-by: Al Finkelstein <al.finkelstein@ibm.com>

tmishina committed:
chore: format code and data

tmishina committed:
refactor: use Pathlib instead of os.path

tmishina committed:
refatcor: remove meaningless header update

tmishina committed:
refactor: readability improvement

tmishina committed:
refatcor: move private functions to the bottom

tmishina committed:
chore: bump version and updated CHANGES.md

tmishina commented:
@alfinkel I have applied your suggestions, bumped version, and updated CHANGES.md. Please tell me if further changes are required, thanks.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,151 @@
+        """
+        for name in cluster_config.namelist():
+            p = pathlib.Path(name)
+            if p.name.startswith('kube-config'):
```
If the file name is in fact named`kube-config` with an extension like `kube_config.ext` then you should use `if p.stem == 'kube-config':`.  Otherwise what you have is appropriate.  Is the file named `kube_config.<some_extension>` or is there truly a need for the `.startswith`?

tmishina replied:
The name of a target file is `kube-config-XXXX.yml` where `XXX` looks the name of the cluster (a document which explicitly says the trailing `XXX` is the name of the cluster does not exist).  The parent directory of the file only contains two files: `kube-config-XXX.yml` and `ca-XXX.pem`, and therefore I think `.startswith('kube-config')` is enough to distinguish between the kube config file and the CA file.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,151 @@
+                kubeconfig = yaml.safe_load(cluster_config.read(name))
+                usr = kubeconfig['users'][0]['user']
+                cluster_token = usr['auth-provider']['config']['id-token']
+            if p.name.endswith('.pem'):
```
```suggestion
            if p.suffix == '.pem':
```

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,151 @@
+                usr = kubeconfig['users'][0]['user']
+                cluster_token = usr['auth-provider']['config']['id-token']
+            if p.name.endswith('.pem'):
+                t = pathlib.Path(self.tempdir.name)
```
What's the point of doing this?  I believe that this just turns the string represented by `self.tempdir.name` into a `Path` object.  Does the `extract` method on the cluster_config ZipFile object accept a `Path` object?  The docs seem to imply that the `path` argument is meant to be a string.  If it's meant to be a string then I suggest you do away with this unneeded wrapping of the path string as a Path object.

tmishina replied:
Test results showed me that `Path` is also acceptable, but as you mentioned documentation assumes this string.  I changed as you suggested in the comment below.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,151 @@
+            if p.name.endswith('.pem'):
+                t = pathlib.Path(self.tempdir.name)
+                cluster_config.extract(name, path=t)
+                ca_cert_filepath = t / name
+        return cluster_token, ca_cert_filepath
```
Again, this makes `ca_cert_filepath` a `Path` object.  I _think_ what you want is a string here?  Does this work currently?  Seems like this is what you'd want here:

```suggestion
                ca_cert_filepath = str(Path(self.tempdir.name, name))
        return cluster_token, ca_cert_filepath
```

tmishina replied:
changed as you suggested (in addition to the change about `p.prefix`).

alfinkel reviewed (changes requested):
In addition to the pathlib comments below, a minor point would be to consider using [pathlib.PurePath](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath) rather than [pathlb.Path](https://docs.python.org/3/library/pathlib.html#pathlib.Path).  You're not dealing with concrete paths here so `PurePath` is a bit more lightweight than `Path` and can be leveraged instead.

tmishina committed:
chore: simplify code

Co-authored-by: Al Finkelstein <al.finkelstein@ibm.com>

tmishina committed:
chore: explicitly returns string instead of Path

tmishina committed:
Merge branch '9-ibm-cloud-resource-fetcher' of github.com:tmishina/auditree-arboretum into 9-ibm-cloud-resource-fetcher

tmishina committed:
chore: use PurePath instead of Path

tmishina commented:
@alfinkel thank you again and again, I updated the code.
- use `PurePath` instead of `Path`
- use string as a parameter for `zipfile.extract()`
- return string instead of Path in `_get_iks_credentials()`
